### PR TITLE
(fix) create subfolders if not exists & improve exception logging

### PR DESCRIPTION
**Pull request by Cka3o4Huk**
Hi,

This patch fixes file creation if subfolder doesn't exist. In addition, it contains few logging improvements like printing of exception message & stacktrace in case of error.

Please review it!

Thank you!

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -132,6 +132,12 @@
 			<scope>compile</scope>
 		</dependency>
 		<dependency>
+			<groupId>org.apache.commons</groupId>
```
Comment (nobeh) [inline pom.xml]: If you do not mind, the way `commons-lang3` is used, is covered by Google Guava. Can we switch for Guava instead?

File: src/main/java/com/narmnevis/range/io/FilePublisher.java [hunk 2]
```diff
@@ -53,4 +54,8 @@ public void close() throws IOException {
 		writer.close();
 	}
 
+	@Override
+	public String toString() {
+		return getClass().getName() + ":" + ((location !=null ) ? location.getAbsolutePath() : "no file");
```
Comment (nobeh) [inline src/main/java/com/narmnevis/range/io/FilePublisher.java]: - There's a formatting space dropped in `location !=null` after `!=`. 
- Can change `no file` to a better wording as "unknown"  or "NA"?

File: src/main/java/com/narmnevis/range/io/PublisherCollection.java [hunk 3]
```diff
@@ -31,6 +32,7 @@ public void publish(RangeContext context) {
 				publisher.publish(context);
 				logger.info("Publisher {} published context {}", publisher, context);
 			} catch (Exception e) {
+				logger.error("Exception occurred: ", ExceptionUtils.getRootCause(e));
```
Comment (nobeh) [inline src/main/java/com/narmnevis/range/io/PublisherCollection.java]: Refer to my comment on Guava, please. The same can be done in Guava by `Throwables` class.

**Comment by nobeh** [comment]
Thank you for another useful PR.

File: pom.xml (hunk 1 above)
Comment (Cka3o4Huk) [inline pom.xml]: Thank you for note! Done.

File: src/main/java/com/narmnevis/range/io/FilePublisher.java (hunk 2 above)
Comment (Cka3o4Huk) [inline src/main/java/com/narmnevis/range/io/FilePublisher.java]: Corrected.

File: src/main/java/com/narmnevis/range/io/PublisherCollection.java (hunk 3 above)
Comment (Cka3o4Huk) [inline src/main/java/com/narmnevis/range/io/PublisherCollection.java]: Done

**Comment by nobeh** [comment]
Thanks for the PR.

**Pull request merged**